src/main/java/edu/harvard/iq/dataverse/FileDownloadServiceBean.java

```diff
@@ -572,5 +572,48 @@ public String getDirectStorageLocatrion(String storageLocation) {
             
         return null; 
     }
-    
+
+    /**
+     * Checks if a user can download a file based on the file metadata and the permissions of the user
+     *
+     * This method is based on {@link edu.harvard.iq.dataverse.FileDownloadHelper#canDownloadFile(FileMetadata),
+     * and has been adapted to make it callable from the API instead of a view
```

Comment:
I developed this endpoint looking for it to execute a logic that simply informs if a file is downloadable by a user, which is a requirement I found necessary for the UI (Take a look at my note about this endpoint in the PR description). The method you mention is executed implicitly when the user performs certain actions on files.

Actually, the logic is not fully copied. In the original method, a fileDownloadPermissionMap is used to serve as a cache. I decided not to go too deep into a refactoring to remove duplication since if I'm not wrong the view scoped classes like the FileDownloadHelper will be removed along with JSF. On the other hand, the absence of tests in the current UI makes any refactoring risky.

Reply:
I'm looking into this now.

Reply:
We do have file access authorization logic duplicated in the 2 places mentioned. The method in Access.java checks the permissions at the time of download. It was however necessary to separately implement the logic in FileDownloadHelper, which is a ViewScoped bean so that the boolean response could be cached the first time it is checked for a specific file. The main reason why the caching was necessary is that by nature of JSF, rendering logic rules always get executed multiple times. It could have been achieved without duplicating all of the logic (there is a comment in Access.java suggesting that we try to consolidate it in one place). But there are some subtle differences between the 2 places to kind of justify it. Plus, as an added secondary benefit, it is possible to make the consecutive permission checks on different files slightly faster as well. Because we can cache other things, for example, whether the user has a certain permission on the dataset level as well (but this is not very significant). 
The "multiple lookups" issue becomes moot in the SPA case (it is within our control to ensure that the javascript does not make unnecessary calls on the same files). For the purposes of creating this new API for checking the permission by the SPA it probably still made sense to base it on this FileDownloadHelper code; to ensure that the logic is exactly the same as what the primefaces page is using; AND because the check is performed on a FileMetadata (and not on a DataFile like in the method in Access.java). But, having stripped everything related to caching (the fileDownloadPermissionMap etc.), the logic left  in the method should indeed match what we have in Access.isAccessAuthorized() 1:1. So, having this method in FileDownloadServiceBean is the first step in switching Access.java to calling it as well. We'll need to double-check that the logic is indeed the same; and it's not something that needs to be done in this PR. 

It's also worth pointing out, that making a logic mistake here is relatively low-risk. The worst that can happen is we give the user a download link and they get a 403 when they click it.

Reply:
Thanks for that great explanation.

Once the caching logic has been stripped, we can also strip the logic that checks the different states of the file (embargoed, deaccessioned, restricted...), since these properties come from the main endpoint and some of them are decisive for the file not to be downloadable. Therefore, we can avoid API calls if these checks are done before on the client side. With all of this logic stripped (cache + statuses checks) the endpoint ends up being a permissions check, called from the SPA when the necessary conditions are met. 

See my other comment on this: https://github.com/IQSS/dataverse/pull/9693#discussion_r1280414517

And the new "permissions only" endpoint: https://github.com/IQSS/dataverse/pull/9693/commits/b52b33dc58962a8356892c497b186dccf63dd9d3